escrow/control_service.go

```diff
@@ -0,0 +1,315 @@
+//One way to determine this is by checking the nonce in the block chain with the nonce in the payment,
+//for a given channel if the block chain nonce is greater than that of the nonce from etcd storage => that the claim is already done in block chain.
+//and the Finish method is called on the claim.
+func (service *ProviderControlService) removeClaimedPayments() error {
+	//Get the pending claims
+	//retrieve all the claims in progress
+	claimsRetrieved, err := service.channelService.ListClaims()
+	if err != nil {
+		return errors.New("error in retrieving claims")
+	}
+	for _, claimRetrieved := range claimsRetrieved {
+		payment := claimRetrieved.Payment()
+		blockChainChannel, ok, err := service.channelService.PaymentChannelFromBlockChain(&PaymentChannelKey{ID: payment.ChannelID})
+		if !ok || err != nil {
+			return err
+		}
+		//Compare the state of this payment in progress with what is available in block chain
+		if blockChainChannel.Nonce.Cmp(payment.ChannelNonce) > 0 {
+			//if the Nonce on this block chain is higher than that of the Payment,
+			//means that the payment has been completed , hence update the etcd state with this
+			log.Debugf("for channel id:%v the nonce of channel from Block chain = %v is "+
+				"greater than nonce of channel from etcd storage :%v , Nonce:%v",
+				payment.ChannelID, blockChainChannel.Nonce, payment.ChannelNonce)
+			err = claimRetrieved.Finish()
+			if err != nil {
+				log.Error(err)
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+//Check if the payment address/signer passed matches to what is present in the metadata
+func (service *ProviderControlService) verifyPaymentAddress(address common.Address) (error) {
+	isSameAddress := service.serviceMetaData.GetPaymentAddress() == address
+	if !isSameAddress {
+		return fmt.Errorf("the payment Address: %s  does not match to what has been registered", blockchain.AddressToHex(&address))
+	}
+	return nil
+}
+
+//Check if the mpe address passed matches to what is present in the metadata.
+func (service *ProviderControlService) checkMpeAddress(mpeAddress string) (error) {
+	isSameAddress := strings.Compare(service.serviceMetaData.MpeAddress, mpeAddress) == 0
+	if !isSameAddress {
+		return fmt.Errorf("the mpeAddress: %s passed does not match to what has been registered", mpeAddress)
+	}
+	return nil
+}
+
+//Check if the block number passed is not more +- 5 from the latest block number on chain
+func compareWithLatestBlockNumber(blockNumberPassed *big.Int) error {
+	latestBlockNumber, err := currentBlock()
+	if err != nil {
+		return err
+	}
+	differenceInBlockNumber := latestBlockNumber.Cmp(blockNumberPassed)
+	if -5 <= differenceInBlockNumber || differenceInBlockNumber <= 5 {
+		return nil
+	}
+	return fmt.Errorf("difference between the latest block chain number and the block number passed is %v ", blockNumberPassed)
+}
+
+//Get the current block number from on chain
+func currentBlock() (*big.Int, error) {
```

Comment:
I think this method can be replaced by `Processor.CurrentBlock()` see https://github.com/singnet/snet-daemon/blob/05b4021d74e0516191289d3b95f68d1f94132385/blockchain/blockchain.go#L114

Reply:
Thanks @vsbogd , was looking at making the  CurrentBlock  a Function rather than a Method ? this way we dont need an instance of *Processor.

Reply:
thanks @vsbogd , will track this as a separately [#189](https://github.com/singnet/snet-daemon/issues/189)